src/Fees.sol

```diff
@@ -1,58 +1,69 @@
-    uint256 constant FIL_USD_DENOM = 35;
-           
-    uint256 constant ONE_PERCENT_DENOM = 100;
-
-    uint256 constant LEAVES_PER_TIB = 2^35;
-
-    uint256 constant EPOCHS_PER_MONTH = 86_400;
-    // Revenue in usd per TiB per month
-    uint256 constant MONTHLY_TIB_REVENUE_USD_NUM = 35;
-    uint256 constant MONTHLY_TIB_REVENUE_USD_DENOM = 10;
-
-    uint256 constant PROOF_GAS_FLOOR = 2_000_000;
-    uint256 constant ONE_NANO_FIL = 10^9;
-    uint256 constant ONE_FIL = 10^18;
-
-    uint256 constant SYBIL_FEE = 10^17;
-
-    // Currently unused
-    //
-    // Returns the service fee for a given proofset duration, and size
-    // Size measured in leaves (32 byte chunks), duration measured in epochs
-    function serviceFee(uint256 duration, uint256 size) internal pure returns (uint256) {
-        // 1 FIL / TiB / month ~ revenue 
-        // 86,400 epochs / month
-        // 2^35 leafs / TiB
-        // fee is 1% of revenue
-
-        // [(Monthly revenue in USD) * (FIL / USD) * size * duration] / 
-        // [Leaves per TiB * Epochs per month * 1%]
-        uint256 numerator = ONE_FIL * MONTHLY_TIB_REVENUE_USD_NUM * FIL_USD_NUM * size * duration;
-        uint256 denominator = MONTHLY_TIB_REVENUE_USD_DENOM * FIL_USD_DENOM * LEAVES_PER_TIB * EPOCHS_PER_MONTH * ONE_PERCENT_DENOM;
-        return numerator / denominator;
-    }
+    // 0.1 FIL
+    uint256 constant SYBIL_FEE = FIL_TO_ATTO_FIL / 10;
+
+    // 2 USD/Tib/month is the current reward earned by Storage Providers
+    uint256 constant MONTHLY_TIB_STORAGE_REWARD_USD = 2;
+    // 1% of reward per period
+    uint256 constant PROOF_FEE_PERCENTAGE = 1;
+    // 4% of reward per period for gas limit left bound
+    uint256 constant GAS_LIMIT_LEFT_PERCENTAGE = 4;
+    // 5% of reward per period for gas limit right bound
+    uint256 constant GAS_LIMIT_RIGHT_PERCENTAGE = 5;
+    uint256 constant USD_DECIMALS = 1e18;
+
+    // 1 TiB in bytes (2^40)
+    uint256 constant TIB_IN_BYTES = 2 ** 40;
+    // Number of epochs per month (30 days * 2880 epochs per day)
+    uint256 constant EPOCHS_PER_MONTH = 86400;
+
+    /// @return proof fee in AttoFIL
+    function proofFeeWithGasFeeBound(
+        uint256 estimatedGasFee, // in AttoFIL
+        uint64 filUsdPrice,
+        int32 filUsdPriceExpo,
+        uint256 rawSize,
+        uint256 nProofEpochs
+    ) internal pure returns (uint256) {
+        require(estimatedGasFee > 0, "failed to validate: estimated gas fee must be greater than 0");
+        require(filUsdPrice > 0, "failed to validate: AttoFIL price must be greater than 0");
+        require(rawSize > 0, "failed to validate: raw size must be greater than 0");
+
+        // Calculate reward per epoch per byte (in AttoFIL)
+        uint256 rewardPerEpochPerByte;
+        if (filUsdPriceExpo >= 0) {
+            rewardPerEpochPerByte = (MONTHLY_TIB_STORAGE_REWARD_USD * FIL_TO_ATTO_FIL) / (TIB_IN_BYTES * EPOCHS_PER_MONTH * filUsdPrice * (10 ** uint32(filUsdPriceExpo)));
```

Comment:
Does pyth oracale have any notion of precision?  The way things are structured here we appear to lose precision exponentially as price increases.  Assuming filUsdPrice is an integer that needs not precision adjustments as we use it here and the pyth oracle can give us FIL price of $1,$2,$3,...,$9, $10, $20, $30,... $100, $200, $300.  If this is the best we can do it is massively better than just estimating with a constant. But ideally we could get constant precision for any price.

Reply:
Ah ok, looking at the docs I think it will typically return a negative exponent and an integer containing lots of precision.  So in this case say FIL = 6.886768 then fiprice will be 6886768 and exponent will be -6 to move the decimal to the correct place.  So in this case the precision will be accounted for.  

Is this how you've seen it work @aarshkshah1992 ?

Reply:
Yes what you are saying is exactly correct. So we need to handle both positive and negative exponents here.

Reply:
I think I was just confusing myself imagining price being between 1 and 10 for some reason.  You can make the price be 133 or 16666 just using the price integer.  I'd be curious to see how much precision the oracle gives and what representation they typically choose.  But I'm not concerned about this anymore.

Reply:
@ZenGround0 I think the exponent is more useful for prices < 1 USD 😅